src/ott/experimental/mmsinkhorn.py

```diff
@@ -0,0 +1,407 @@
+
+  def _diverged(self, state: MMSinkhornState, iteration: int) -> bool:
+    err = state.errors[iteration // self.inner_iterations - 1, 0]
+    return jnp.logical_not(jnp.isfinite(err))
+
+  def _continue(self, state: MMSinkhornState, iteration: int) -> bool:
+    """Continue while not(converged) and not(diverged)."""
+    return jnp.logical_and(
+        jnp.logical_not(self._diverged(state, iteration)),
+        jnp.logical_not(self._converged(state, iteration))
+    )
+
+  @property
+  def outer_iterations(self) -> int:
+    """Upper bound on number of times inner_iterations are carried out.
+
+    This integer can be used to set constant array sizes to track the algorithm
+    progress, notably errors.
+    """
+    return np.ceil(self.max_iterations / self.inner_iterations).astype(int)
+
+  def tree_flatten(self):  # noqa: D102
+    aux = vars(self).copy()
+    aux.pop("threshold")
+    return [self.threshold], aux
+
+  @classmethod
+  def tree_unflatten(cls, aux_data, children):  # noqa: D102
+    return cls(**aux_data, threshold=children[0])
+
+
+def run(const, solver, state):
+  cost_t, a_s, epsilon = const
+
+  def cond_fn(iteration: int, const: Any, state: MMSinkhornState) -> bool:
+    return solver._continue(state, iteration)
+
+  def body_fn(
+      iteration: int, const: Tuple[jnp.ndarray, ...],
+      state: Tuple[jnp.ndarray, ...], compute_error: bool
+  ) -> Tuple[jnp.ndarray, float]:
+    cost_t, a_s, epsilon = const
+    k = len(a_s)
+
+    def one_slice(potentials: Tuple[jnp.ndarray, ...], l: int, a: jnp.ndarray):
+      k = len(potentials)
+      pot = potentials[l]
+      axis = list(range(l)) + list(range(l + 1, k))
+      pot += epsilon * jnp.log(a) + mu.softmin(
+          remove_tensor_sum(cost_t, potentials), epsilon, axis=axis
+      )
+      return potentials[:l] + [pot] + potentials[l + 1:]
+
+    potentials = state.potentials
+    for l in range(k):
+      potentials = jax.jit(
+          one_slice, static_argnames="l"
+      )(potentials, l, a_s[l])
+
+    state = state.set(potentials=potentials)
+    err = jax.lax.cond(
+        jnp.logical_or(
+            iteration == solver.max_iterations - 1,
+            jnp.logical_and(compute_error, iteration >= solver.min_iterations)
+        ),
+        lambda state, c, a, e: state.solution_error(c, a, e, solver.norm_error),
+        lambda *_: jnp.inf, state, cost_t, a_s, epsilon
+    )
+    errors = state.errors.at[iteration // solver.inner_iterations, :].set(err)
+    return state.set(errors=errors)
+
+  fix_point = fixed_point_loop.fixpoint_iter_backprop
+  const = (cost_t, a_s, epsilon)
+  state = fix_point(
+      cond_fn, body_fn, solver.min_iterations, solver.max_iterations,
+      solver.inner_iterations, const, state
+  )
+  converged = jnp.logical_and(
+      jnp.logical_not(jnp.any(jnp.isnan(state.errors))), state.errors[-1]
+      < solver.threshold
+  )[0]
+  # Compute cost
+  out = MMSinkhornOutput(
```

Comment:
This follows mostly the `Sinkhorn` template in which we don't materialize the cost in the output. In the absence of a suitable `MMGeometry` or a `MMPointCloud` I thought this would be a suitable compromise.